Widget class

vcoindet opened the pull request:
Adding Widget class in example

Widget class with constructor and method (Widget.js)

In Widget.js
- getName()
- getElement()
- getOptions()
- setOptions(pOptions)
- listenToMap(pEventName, pCallback)
- getGlobe()
- setGlobe(map)

Method in widgetAPI.js
- addWidget(widget)
- getWidgets()
- removeWidget(widget)

To make the examples work : put the following lines in index.html :
- < link rel="stylesheet" href="examples/widgets/widget.css" > in the 'head' tag
- <script src="examples/widgets/pickPositionWidget.js"> </script>
  <script src="examples/widgets/rangeWidget.js"> </script> 
  Right before the closure of the 'body' tag
- <script src="examples/widgets/Widget.js"> </script>
  <script src="examples/widgets/widgetAPI.js"> </script>
  Just after the line <script src="dist/itowns.js"></script>
- const widgetTab = new widgetAPI();

vcoindet commented:
rebased and ready

autra commented on examples/controls/control.css:
```diff
@@ -0,0 +1,12 @@
+      top:10px;
+      right:10px;
+      width:150px;
+      height:150px;
```
nits: space before `:`

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -852,6 +857,45 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
 
+/**
+ * Add a control.
+ * @constructor
```
Not a constructor! (and below too)

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -852,6 +857,45 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
+ */
+
+ApiGlobe.prototype.addControl = function addControl(control) {
+    var index = this.controls.indexOf(control);
```
Maybe a Set is a more natural data structure here.

vcoindet replied:
You're probably right, didn't know this method at the time. Will change that. Or Maybe a `new Map()` can do the job too.

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+/**
+ * Generated On: 2016-11-22
+ * Class: Controls
+ * Description: Cette Classe construit un control.
```
Should be in English

peppsac commented:
@vcoindet can you please squash or reword your commits? (5 commits use the same text)

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+    this.api = ApiGlobe;
+}
+
+Control.prototype.constructor = Control;
```
you don't need this if you don't overwrite the `prototype`.

vcoindet replied:
OK

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+    this.pName = pName;
+    this.pElement = pElement;
+    this.options = options;
+    this.map_ = null;
```
Is it supposed to be a private field? In this case, the convention is to name it `this._map`.

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+    document.getElementById('viewerDiv').addEventListener(pEventName, pCallback, false);
+};
+
+Control.prototype.getGlobe = function getGlobe() {
```
Might not be a globe in all cases (we want to make the geometry configurable in the future right?)

autra replied:
`getMap` and `setMap`?

vcoindet replied:
Yep, I guess you're right, I hesitated between those two choices, but I only thought about 3D and not 2D

autra commented on src/Core/Commander/Interfaces/Control.js:
```diff
@@ -0,0 +1,67 @@
+/**
```
One of `src/Core/Commander/Interfaces/Control.js` and `src/Core/Commander/Interfaces/Control/Control.js` should be removed I guess.

vcoindet replied:
Oups, my mistake, the first one has to be removed...

autra commented on index.html:
```diff
@@ -42,6 +53,7 @@
             // iTowns namespace defined here
             var viewerDiv = document.getElementById('viewerDiv');
             var menuGlobe = new GuiTools(itowns.viewer,'menuDiv');
+
```
Not very critical, but in general it's best to remove unrelated changes from a PR. (And submit them elsewhere if needed).

autra commented on index.html:
```diff
@@ -68,7 +80,7 @@
             Promise.all([imageryPromise, elevationPromise]).then(() => itowns.viewer.init());
 
-            menuGlobe.addGUI('RealisticLighting',false,function(newValue) {itowns.viewer.setRealisticLightingOn(newValue);});
+            menuGlobe.addGUI('RealisticLighting',false,(newValue) => {itowns.viewer.setRealisticLightingOn(newValue);});
```
Same remark, unrelated change.

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -660,6 +663,7 @@ ApiGlobe.prototype.setRange = function setRange(pRange /* , bool anim*/) {
     eventRange.oldRange = this.getRange();
     loaded = false;
     this.scene.currentControls().setRange(pRange);
+    this.viewerDiv.dispatchEvent(eventRange);
```
Be careful, events should not be used as static variable. THe browser sets some fields (`targets`, `innerTargets`). By using the same ref for every events, you risk some collision. It's best to create a new instance at each `dispatch`.

vcoindet replied:
So I guess we should change that for all dispatched events. Maybe in an another PR ? WDYT ?

autra replied:
yep, you're right :-)

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+
+Control.prototype.setGlobe = function setGlobe(map) {
+    if (this.map_) {
+        document.getElementById(this.options.div.id).removeChild(this.pElement);
```
If I'm not mistaken `this.options.div` is already the same dom element returned by `document.getElementById(this.options.div.id)`.

vcoindet replied:
Well, I'm not sure of that. I have to test it, I will let you know

vcoindet replied:
Yep, you're right, so I changed it

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+    this.map_ = map;
+    if (this.map_) {
+        var element = this.pElement;
+        document.getElementById(this.options.div.id).appendChild(element);
```
same here.

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+ */
+
+Control.prototype.getOptions = function getOptions() {
+    this.options = this.options || '';
```
you meant `this.options = this.options || {};` right?

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+ */
+
+Control.prototype.listenToMap = function listenToMap(pEventName, pCallback) {
+    document.getElementById('viewerDiv').addEventListener(pEventName, pCallback, false);
```
there should be a `removeEventListener` somewhere, as apparently you can remove controls. Failure to do so would prevent the browser to dispose of your object.

autra commented on examples/controls/control.css:
```diff
@@ -0,0 +1,12 @@
+      top:10px;
+      right:10px;
+      width:150px;
+      height:150px;
```
Consider using `rem` units instead of px to accomodate with different screen size and density.

vcoindet replied:
Will work on that

autra commented on examples/controls/control.css:
```diff
@@ -0,0 +1,12 @@
+      border-style: solid;
+      border-color: black;
+      border-width: 1px;
+      box-shadow: 1px 1px 10px #555;
```
(for border and boxes, some people prefer staying with px, they think it's nicer)

autra commented on examples/controls/control.css:
```diff
@@ -0,0 +1,12 @@
+#controlDiv {
```
This file should be called `controls.css`, and this div `controlsDiv`. Or even better `#controls`. The fact it is a div is an implementation detail.

vcoindet replied:
OK I agree

autra commented on examples/controls/control.css:
```diff
@@ -0,0 +1,12 @@
+      border-color: black;
+      border-width: 1px;
+      box-shadow: 1px 1px 10px #555;
+}
```
I actually suggest:

```css
position: absolute;
top: 1rem;
right: 1rem;
width: 20rem;
box-shadow: 1px 1px 10px #555;
font: 11px Lucida Grande,sans-serif;
color: #eee;
padding: 0.5rem;
```

And separating the fields with `<br>` instead of `;` in your callbacks (in `pickposition_controls`).

Result:

![screen shot 2017-01-16 at 15 22 27](https://cloud.githubusercontent.com/assets/1576598/21986416/a69406fa-dbff-11e6-989c-6bc3355eccfa.png)

vcoindet replied:
I don't usually use `<br>` for semantic reasons. I'd rather use `<li>`. I agree that `;` wasn't a good solution, but it was just an example to test the control class

autra replied:
yep :-) Indeed, `li` are even better :-)

autra commented on examples/controls/pickPosition_control.js:
```diff
@@ -0,0 +1,21 @@
+var pickControl = document.createElement('div');
```
For the filename: either choose snake_case or CamelCase. (the rest of the project use CamelCase if I'm not mistaken).

vcoindet replied:
lowerCamelCase if I'm not mistaken too. The best is that i choose lowerCamelCase.

autra replied:
Actually yes, lowerCamelCase, thanks for translating for me ;-)

autra commented on examples/controls/pickPosition_control.js:
```diff
@@ -0,0 +1,21 @@
+    'Pick',
+    document.getElementById('pickControl'),
+    {
+        div: document.getElementById('controlDiv')
```
`div: controlDiv`. You have it just above.

vcoindet replied:
absolutely !

autra commented on examples/controls/pickPosition_control.js:
```diff
@@ -0,0 +1,21 @@
+    document.getElementById(control.pElement.id).innerHTML = "longitude : " + itowns.viewer.pickPosition().coordinate[0] * 180 / 3.14 + " ; latitude : " + itowns.viewer.pickPosition().coordinate[1] * 180 / 3.14 + " ; altitude : " +  itowns.viewer.pickPosition().coordinate[2];
+};
+
+itowns.viewer.addEventListener('mousedown', callback);
```
So I don't understand: how is `Control` useful if you do all the logic outside of it? You're not even using `listenToMap`... (though I think a callback in the `options` constructor parameter would be a better idea than listenToMap).

autra commented on src/Core/Commander/Interfaces/ApiInterface/ApiGlobe.js:
```diff
@@ -852,6 +857,45 @@ ApiGlobe.prototype.removeEventListenerLayerChanged = function removeEventListene
+    if (index > -1) {
+        this.controls.splice(index, 1);
+    }
+    control.setGlobe();
```
Let's be explicit `control.setGlobe(null);`

But it sounds like an abuse to me. I would prefer having a `control.dispose` method that would remove listeners and stuff.

autra commented on index.html:
```diff
@@ -68,7 +80,7 @@
+            menuGlobe.addGUI('RealisticLighting',false,(newValue) => {itowns.viewer.setRealisticLightingOn(newValue);});
 
            	itowns.viewer.addEventListener('globe-loaded', function() {
             	console.log("Globe Loaded");
```
Actually, we could have several example html, one basic like it is now, another with all the controls etc... WDYT?

vcoindet replied:
I totally agree, as controls like the menu, IPR, range, picking, etc... should be examples.

autra commented on src/Core/Commander/Interfaces/Control/Control.js:
```diff
@@ -0,0 +1,88 @@
+ * Description: Cette Classe construit un control.
+ */
+
+import ApiGlobe from 'Core/Commander/Interfaces/ApiInterface/ApiGlobe';
```
Not used anywhere. Also, cyclic dependencies should be avoided as much as possible (it creates difficult-to-maintain code).

vcoindet replied:
You're absolutely right, I forget to remove it

autra reviewed (commented):
As nobody was looking at your PR, I took upon myself to do it :-) 

Can you explain to me what is the exact role of `Controls`? Is it to display information from some events? if so, maybe we can find another name (controls is for controlling stuff. We're not really controlling them here). `indicators`? `sensors`?

Also, my general feeling about this is that it is quite a rigid way to give information to client code. Don't we just want some events (optionally embedding useful informations, like new values) and let client code (geoportal, anything) deal with that how they see fit? 

As it is now, if I just want to store range info somewhere (for history or telemetry purpose for example), I cannot do that without displaying them. But maybe I didn't get the rationale here.

vcoindet commented:
To be clear on the purpose of the Controls class : 
A control is a widget with a DOM element which position should be fixed on the screen (determined by using CSS) . It can involve inputs or be informational only.
You can use it for simple custom controls by creating the element with listeners, for example displaying the IPRs on the screen : IPRs can change when you remove a layer (event `layerchanged:visible`) or when you change the center of the map.

Here the control class is inspired on the OpenLayers 3's control class.

All the methods implemented in the Controls class can be used to help you display your control on the screen. So basically, it is used to create DOM elements and all controls inherits of this class.
Some examples of controls : menu, IPR, scale bar, mini Globe, zoom and dezoom button, etc... (see the screenshot for IPRs below)
![ipr](https://cloud.githubusercontent.com/assets/16044646/22198833/92df0dfe-e158-11e6-8137-40d56f06bbd0.png)

autra commented:
Thanks for the context !

So actually, those Controls would exactly do dat.gui's job + displaying read-only informations, right? If I understand it correctly, why don't we use dat.js for this? It's already a very good state-of-the-art lib, with more than 1000 stars on github... and we are already using it! So it would mean less maintenance for us, and a better integration with the currently enabled controls. 

For example, with the current master, with a small modification to GuiTools.js (return the result of `this.gui..add` on line 74), I can achieve a range spy just with these following lines in index.html:
```js
let myUi = menuGlobe.addGUI('range', {itowns.viewer.getRange()}, (newValue) => { /* do nothing */ });
myUi.domElement.children[0].disabled = true; // make this read-only
itowns.viewer.addEventListener('rangeChanged', () => {
    console.log('new range', itowns.viewer.getRange());
    menuGlobe.range = itowns.viewer.getRange();
    myUi.updateDisplay();
});
```
So with some polishing (creating a new method in GuiTools or modifiying `addGUI` by adding a `readonly` boolean parameter and an event name to listen to), we already have everything we need, no need to create another module for that (+ some css etc).

Note that the example above only works if you manually execute `itowns.viewer.setRange(...)` in the console, but that's because we don't correctly fire `rangechanged` events when we do it from GlobeControls.js. So this needs to get fixed anyway (rangechanged events should be fired from GlobeControls I guess).

So to summarize, I thing that:
- we should fix our events in ApiGlobe and ensure they are fired when they should.
- we should use dat.gui for these controls.

One last thing:

> A control is a widget

So let's call them Widgets? Because for us, controls are a different thing (mouse and keyboard actions), ex GlobeControls.js (and I hope to make them configurable and to have FPSControls, PanoramicControls etc...).

vcoindet commented:
Commit squashed, Rebased and Ready
Changed the words 'controls' by 'widgets' as suggested by @autra

vcoindet committed:
Adding Widget class in example

Widget class with constructor and method (Widget.js)

In Widget.js

- getName()
- getElement()
- getOptions()
- setOptions(pOptions)
- listenToMap(pEventName, pCallback)
- getGlobe()
- setGlobe(map)

Method in widgetAPI.js

- addWidget(widget)
- getWidgets()
- removeWidget(widget)

To make the examples work : put the following lines in index.html :

< link rel="stylesheet" href="examples/widgets/widget.css" > in the 'head' tag
<script src="examples/widgetss/pickPositionWidget.js"> </script> <script src="examples/widgets/rangeWidget.js"> </script> Right before the closure of the 'body' tag
<script src="examples/widgets/Widget.js"> </script> <script src="examples/widgets/widgetAPI.js"> </script> Just after the line <script src="dist/itowns.js"></script>
const widgetTab = new widgetAPI();

vcoindet commented:
Rebased and Ready

Put everything in examples, please read the description of this PR to make it work correctly.

gchoqueux commented on .gitattributes:
```diff
@@ -1,4 +1,4 @@
-*	text=auto
+*   text=auto
```
reason?

gchoqueux commented on examples/widgets/Widget.js:
```diff
@@ -0,0 +1,104 @@
+
+Widget.prototype.setMap = function setMap(map) {
+    if (this._map) {
+        this.options.div.removeChild(this.pElement);
```
why remove and re-add this.pElement?

gchoqueux commented on examples/widgets/Widget.js:
```diff
@@ -0,0 +1,104 @@
+    }
+    this._map = map;
+    if (this._map) {
+        var element = this.pElement;
```
remove `var element` and replace `this.options.div.appendChild(element);` by `this.options.div.appendChild(this.pElement);`

gchoqueux commented on examples/widgets/pickPositionWidget.js:
```diff
@@ -0,0 +1,37 @@
+/* global itowns */
```
rename by PickPositionWidget.js

gchoqueux commented on examples/widgets/Widget.js:
```diff
@@ -0,0 +1,104 @@
+ * Description: Create a widget.
+ */
+
+function Widget(pName, pElement, options) {
```
export Widget

autra replied:
I might be late to the party, but you cannot really do that: the examples are neither handled by webpack nor by babel. Imho they should be seen as ready-to-use snippets of code for client apps. Therefore, they shouldn't rely on webpack, babel or anything linked to our build system. They should be valid, browser-ready javascript for our target browsers.

This actually means that:
- we need to formally decide which our target browsers (and versions) are (because it defines which js api/functions/syntax we can safely use)
- we need a special `.eslintrc` for the `examples` folder

vcoindet replied:
Sorry for the late answer... You're not too late ! As you mentionned it, it's not possible to do that in `examples`.
In fact, I thought maybe we could do a new `index.html` in `/examples/` for each widget's examples, but it won't solve the problem.
I agree with your solution, I think it's the best we can think about right now. But I don't have any knowledge on `.eslintrc`, so do you agree to help me to do a special `.eslintrc` for the `examples` folder ?

gchoqueux commented on examples/widgets/pickPositionWidget.js:
```diff
@@ -0,0 +1,37 @@
+/* global itowns */
+/* global Widget */
```
import Widget

gchoqueux commented on examples/widgets/pickPositionWidget.js:
```diff
@@ -0,0 +1,37 @@
+/* global itowns */
+/* global Widget */
+/* global widgetTab */
```
remove global  widgetTab 

User must do widgetTab.add(pickPositionWidget);

Do example to test pickPositionWidget

gchoqueux commented on examples/widgets/pickPositionWidget.js:
```diff
@@ -0,0 +1,37 @@
+        document.getElementById(widget.pElement.id).removeChild(document.getElementById('pickList'));
+    }
+    const tab = ['longitude', 'latitude', 'altitude'];
+    const coords = itowns.viewer.pickPosition().coordinate;
```
remove global api call, pass in api in constructor

gchoqueux commented on examples/widgets/rangeWidget.js:
```diff
@@ -0,0 +1,31 @@
+/* global itowns */
```
no global variable and use import export, see comment for pick

rename file by RangeWidget.js

gchoqueux replied:
add example

gchoqueux commented on examples/widgets/widgetAPI.js:
```diff
@@ -0,0 +1,34 @@
+/* global Widget */
+/* global itowns */
+
+function widgetAPI() {
```
widgetApi is not really necessary

gchoqueux commented on examples/widgets/pickPositionWidget.js:
```diff
@@ -0,0 +1,37 @@
+    const coords = itowns.viewer.pickPosition().coordinate;
+    const list = document.createElement('ul');
+    list.id = 'pickList';
+    for (var j = 0; j < coords.length; j++) {
```
replace by for ... in

gchoqueux replied:
sorry replace by for ... of

autra replied:
(if for..of is supported by our target browsers, we won't use babel here I guess)
